src/lib/pickles/inductive_rule.mli

```diff
@@ -1,152 +1,133 @@
-        previous proofs where the first had 2 inductive rules and the second
-        had 1.
-    - ['a_var] is the in-circuit type of the [main] function's public input.
-    - ['a_value] is the out-of-circuit type of the [main] function's public
-      input.
-    - ['ret_var] is the in-circuit type of the [main] function's public output.
-    - ['ret_value] is the out-of-circuit type of the [main] function's public
-      output.
-    - ['auxiliary_var] is the in-circuit type of the [main] function's
-      auxiliary data, to be returned to the prover but not exposed in the
-      public input.
-    - ['auxiliary_value] is the out-of-circuit type of the [main] function's
-      auxiliary data, to be returned to the prover but not exposed in the
-      public input.
-  *)
-  type ( 'prev_vars
-       , 'prev_values
-       , 'widths
-       , 'heights
-       , 'a_var
-       , 'a_value
-       , 'ret_var
-       , 'ret_value
-       , 'auxiliary_var
-       , 'auxiliary_value )
-       t =
-    { identifier : string
-    ; prevs :
+    module T
+        (Statement : Pickles_types.Poly_types.T0)
+        (Statement_value : Pickles_types.Poly_types.T0)
+        (Return_var : Pickles_types.Poly_types.T0)
+        (Return_value : Pickles_types.Poly_types.T0)
+        (Auxiliary_var : Pickles_types.Poly_types.T0)
+        (Auxiliary_value : Pickles_types.Poly_types.T0) : sig
+      type nonrec ('prev_vars, 'prev_values, 'widths, 'heights) t =
         ( 'prev_vars
         , 'prev_values
         , 'widths
-        , 'heights )
-        Pickles_types.Hlist.H4.T(Tag).t
-    ; main :
-           'a_var main_input
-        -> ('prev_vars, 'widths, 'ret_var, 'auxiliary_var) main_return M.t
-    ; feature_flags : bool Pickles_types.Plonk_types.Features.t
-    }
+        , 'heights
+        , Statement.t
+        , Statement_value.t
+        , Return_var.t
+        , Return_value.t
+        , Auxiliary_var.t
+        , Auxiliary_value.t )
+        t
+    end
+  end
 
-  module T
-      (Statement : Pickles_types.Poly_types.T0)
-      (Statement_value : Pickles_types.Poly_types.T0)
-      (Return_var : Pickles_types.Poly_types.T0)
-      (Return_value : Pickles_types.Poly_types.T0)
-      (Auxiliary_var : Pickles_types.Poly_types.T0)
-      (Auxiliary_value : Pickles_types.Poly_types.T0) : sig
-    type nonrec ('prev_vars, 'prev_values, 'widths, 'heights) t =
-      ( 'prev_vars
-      , 'prev_values
-      , 'widths
-      , 'heights
-      , Statement.t
-      , Statement_value.t
-      , Return_var.t
-      , Return_value.t
-      , Auxiliary_var.t
-      , Auxiliary_value.t )
-      t
+  module Promise : sig
+    include module type of Make (Promise)
   end
+
+  module Deferred : sig
+    include module type of Make (Async_kernel.Deferred)
+  end
+
+  include module type of Make (Pickles_types.Hlist.Id)
 end
 
-module Promise : sig
-  include module type of Make (Promise)
+module type Proof_intf = sig
```

Comment:
Wondering how much this conflicts with `proof.ml`/proof.mli`, and if we could not merge/move some logic there.
It seems that Kimchi is instantiated with `Proof` below, which implements `Proof_intf`, so I think there is definitely some mixed logic.

Reply:
The ultimate goal is to be able to instantiate a  kimchi and saffron module, i.e 

```ocaml
(*this is what we currently have *)
module Kimchi = Make (Proof)

(* coming soon *)
module Saffron = Make (Saffron.Proof)
``` 

We currently don't know what `Proof_intf` will look like so that the rest of pickles will work without using a concrete `Kimchi` module, but with the next iteration we will know more.